PopPUNK/network.py

```diff
@@ -568,30 +587,65 @@ def constructNetwork(rlist, qlist, assignments, within_label,
+            G_df = cudf.DataFrame()
+            G_df['source'] = sparse_input.row
+            G_df['destination'] =  sparse_input.col
+        elif G_df is None:
+            G_df = pd.DataFrame()
+            G_df['source'] = sparse_input.row
+            G_df['destination'] =  sparse_input.col
+        G_df['weights'] = sparse_input.data
     else:
-        for row_idx, (assignment, (ref, query)) in enumerate(zip(assignments,
-                                                                 listDistInts(rlist, qlist,
-                                                                              self = self_comparison))):
-            if assignment == within_label:
-                if weights is not None:
-                    if weights_type == 'euclidean':
-                        dist = np.linalg.norm(weights[row_idx, :])
-                    elif weights_type == 'core':
-                        dist = weights[row_idx, 0]
-                    elif weights_type == 'accessory':
-                        dist = weights[row_idx, 1]
-                    edge_tuple = (ref, query, dist)
-                else:
-                    edge_tuple = (ref, query)
-                connections.append(edge_tuple)
+        if use_gpu and G_df is None:
+            # Add node indices to DF
+            edge_array = cp.array(list(
+                                        listDistInts(rlist,
+                                                     qlist,
+                                                     self = self_comparison)
+                                    ),
+                                    dtype = np.int32)
+            edge_gpu_matrix = cuda.to_device(edge_array)
+            G_df = cudf.DataFrame(edge_gpu_matrix, columns = ['source','destination'])
+        elif G_df is None:
+            # Add node indices to DF
+            G_df = pd.DataFrame(list(listDistInts(rlist, qlist, self = self_comparison)),
+                                columns = ['source','destination'])
+        # Add further information to DF
+        if 'src' in G_df.columns:
+            G_df.rename(columns={'src': 'source', 'dst': 'destination'})
+        G_df['assignments'] = assignments
+        if weights is not None:
+            if weights_type == 'euclidean':
+                G_df['weights'] = np.linalg.norm(weights, axis = 1)
+            elif weights_type == 'core':
+                G_df['weights'] = weights[:, 0]
+            elif weights_type == 'accessory':
+                G_df['weights'] = weights[:, 1]
+
+        # Select rows
+        G_df = G_df[G_df['assignments'] == within_label]
+
+        # Select columns
+        if weights is not None:
+            G_df = G_df[['source','destination','weights']]
+        else:
+            G_df = G_df[['source','destination']]
+
+    if not use_gpu:
+        # Convert to tuples and delete the unnecessary data frame
+        connections = list(zip(*[G_df[c].values.tolist() for c in G_df]))
+        del G_df
```

Comment:
feels wrong to delete an input argument. Better to delete where this is passed?

Reply:
It may just help a lot if you rethink the overall if statement. `if G_df != None` feels like it should be its own branch?

Reply:
It helped in some test runs where GPU memory was exhausted and memory management was not set to share CPU memory - I don't think the marginal gain is worth the potential problems you highlight now memory management is (hopefully still) working. Removed in 8fc9763.